Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java

```diff
@@ -260,94 +241,194 @@ public String getJobUuid() {
-                                    } else {
-                                        return getResponsePayloadWithState()
-                                                .flatMap(responsePayload ->
-                                                        Observable.just(Pair.of(logLines, responsePayload.getState()))
-                                                        .delay(getDelaySeconds(), TimeUnit.SECONDS));
-                                    }
-                                })
-                                .repeatWhen(ob -> ob.delay(GET_LOG_REPEAT_DELAY_MILLISECONDS, TimeUnit.MILLISECONDS))
-                                // FIXME:
-                                // For HDI spark job, the ending condition is `jobState != starting || appIdIsAllocated`
-                                // However, currently the response has no appId all the time so we ignored the app id check
-                                .takeUntil(logAndStatePair ->
-                                        !logAndStatePair.getRight().toString()
-                                                .equalsIgnoreCase(SparkBatchJobState.STARTING.toString()))
-                                .flatMap(logAndStatePair -> Observable.from(logAndStatePair.getLeft()))
-                                .map(line -> new AbstractMap.SimpleImmutableEntry<>(Log, line))
-                                .retryWhen(error -> error.flatMap(exception -> {
-                                    if (exception instanceof SparkJobFinishedException) {
-                                        throw propagate(exception);
-                                    } else {
-                                        // delay sometime and retry with limited times
-                                        return Observable.just(1)
-                                                .zipWith(Observable.range(1, getRetriesMax()), (n, i) -> i)
-                                                .doOnNext(i -> ctrlInfo("Retry retrieving livy submission log..."))
-                                                .delay(getDelaySeconds(), TimeUnit.SECONDS);
-                                    }
-                                }))
+        return getSparkBatchJobRequest()
+                .retryWhen(err ->
+                        err.zipWith(Observable.range(1, getRetriesMax()), (n, i) -> i)
+                                .delay(GET_LIVY_URL_RETRY_DELAY_MILLISECONDS, TimeUnit.MILLISECONDS)
                 )
-                .onErrorResumeNext(err -> {
-                    if (err instanceof SparkJobFinishedException || err.getCause() instanceof SparkJobFinishedException) {
-                        return Observable.error(err);
+                .repeatWhen(ob -> ob.delay(GET_LIVY_URL_REPEAT_DELAY_MILLISECONDS, TimeUnit.MILLISECONDS))
+                .takeUntil(batchResp -> isJobEnded(batchResp) || StringUtils.isNotEmpty(getLivyAPI(batchResp)))
```

Comment:
How about if `getJobState()` always returns `null`?

Reply:
Throw an exception in the new iteration.

---
In reply to: [268477836](https://github.com/Microsoft/azure-tools-for-java/pull/2899#discussion_r268477836) [](ancestors = 268477836)